Userland/Libraries/LibWeb/Painting/BorderPainting.cpp

```diff
@@ -140,69 +159,335 @@ void paint_border(PaintContext& context, BorderEdge edge, DevicePixelRect const&
+            0,
+            false,
+            false);
+
+        // if joined borders have the same color, combine them to draw together
+        if (ready_to_draw) {
+            Gfx::AntiAliasingPainter aa_painter { context.painter() };
+            path.close_all_subpaths();
+            aa_painter.fill_path(path, color, Gfx::Painter::WindingRule::EvenOdd);
+            path.clear();
+        }
     };
 
-    auto draw_border = [&](auto const& border, auto const& radius, auto const& opposite_border, auto const& opposite_radius, auto p1_step_translate, auto p2_step_translate) {
-        auto [p1, p2] = points_for_edge(edge, rect);
-        auto p1_step = radius ? 0 : context.enclosing_device_pixels(border.width).value() / device_pixel_width.value();
-        auto p2_step = opposite_radius ? 0 : context.enclosing_device_pixels(opposite_border.width).value() / device_pixel_width.value();
-        for (DevicePixels i = 0; i < device_pixel_width; ++i) {
-            draw_horizontal_or_vertical_line(p1, p2);
-            p1_step_translate(p1, p1_step);
-            p2_step_translate(p2, p2_step);
+    auto compute_midpoint = [](int horizontal_radius, int vertical_radius) {
+        if (horizontal_radius == 0 && vertical_radius == 0) {
+            return Gfx::FloatPoint(0, 0);
         }
+        // FIXME: this middle point rule seems not exacly the same as main browsers
+        // compute the midpoint based on point whose tangent slope of 1
+        // https://math.stackexchange.com/questions/3325134/find-the-points-on-the-ellipse-where-the-slope-of-the-tangent-line-is-1
+        return Gfx::FloatPoint(
+            (horizontal_radius * horizontal_radius) / AK::sqrt(1.0f * horizontal_radius * horizontal_radius + vertical_radius * vertical_radius),
+            (vertical_radius * vertical_radius) / AK::sqrt(1.0f * horizontal_radius * horizontal_radius + vertical_radius * vertical_radius));
     };
 
-    // FIXME: There is some overlap where two borders (without border radii meet),
-    // which produces artifacts if the border color has some transparency.
-    // (this only happens if the angle between the two borders is not 45 degrees)
+    /**
+     *   0 /-------------\ 7
+     *    / /-----------\ \
+     *   /-/ 3         4 \-\
+     *  1  2             5  6
+     * For each border edge, need to compute 8 points at most, then paint them as closed path.
+     * 8 points are the most complicated case, it happens when the joined border width is not 0 and border radius larger than border width on both side.
+     * If border radius is smaller than the border width, then the inner corner of the border corner is a right angle.
+     */
     switch (edge) {
-    case BorderEdge::Top:
+    case BorderEdge::Top: {
+        auto joined_border_width = context.enclosing_device_pixels(borders_data.left.width);
+        auto opposite_joined_border_width = context.enclosing_device_pixels(borders_data.right.width);
```

Comment:
Would it be possible to share (some of) the code in these cases? Each case looks pretty similar to one another.

Reply:
It seems sharing this part still need another switch case statement. Maybe leave them here is ok.. 🤔